File: Assets/formatting.py
#!/usr/bin/env python3

import re
import os
import sys
import subprocess
from pathlib import Path
from typing import List, Tuple, Optional
from dataclasses import dataclass
from enum import Enum


class ImportType(Enum):
    QT_CORE = 1
    QT_QUICK = 2
    QT_LABS = 3
    QT_QUICK_OTHER = 4
    QUICKSHELL = 5
    QUICKSHELL_OTHER = 6
    QS_COMPONENTS = 7
    EXPLICIT = 8


class PropertyType(Enum):
    ALIAS = 1
    REQUIRED = 2
    READONLY = 3
    NORMAL = 4


@dataclass
class Import:
    line: str
    type: ImportType
    original_index: int


@dataclass
class Property:
    line: str
    type: PropertyType
    indent: str


def find_git_root() -> Optional[Path]:
    """Find the root directory of the git repository."""
    try:
        result = subprocess.run(
            ["git", "rev-parse", "--show-toplevel"],
            capture_output=True,
            text=True,
            check=True,
        )
        return Path(result.stdout.strip())
    except (subprocess.CalledProcessError, FileNotFoundError):
        return None


def classify_import(import_line: str) -> ImportType:
    import_match = re.match(r"import\s+([^\s]+)", import_line)
    if not import_match:
        return ImportType.EXPLICIT

    module = import_match.group(1)

    if module.startswith('"') or module.startswith("'"):
        return ImportType.EXPLICIT

    if module == "QtCore":
        return ImportType.QT_CORE
    elif module == "QtQuick":
        return ImportType.QT_QUICK
    elif module.startswith("Qt.labs."):
        return ImportType.QT_LABS
    elif module.startswith("QtQuick."):
        return ImportType.QT_QUICK_OTHER
    elif module == "Quickshell":
        return ImportType.QUICKSHELL
    elif module.startswith("Quickshell."):
        return ImportType.QUICKSHELL_OTHER
    elif module.startswith("qs."):
        return ImportType.QS_COMPONENTS
    else:
        return ImportType.EXPLICIT


def sort_imports(lines: List[str]) -> Tuple[List[str], bool]:
    imports = []
    non_import_start = 0

    # Get all imports
    for i, line in enumerate(lines):
        stripped = line.strip()
        if stripped.startswith("import "):
            import_type = classify_import(stripped)
            imports.append(Import(line, import_type, i))
        elif stripped and not stripped.startswith("//"):
            non_import_start = i
            break

    if not imports:
        return lines, False

    sorted_imports = sorted(imports, key=lambda x: x.type.value)

    result = []
    prev_type = None

    for imp in sorted_imports:
        current_type = imp.type

        if prev_type is not None:
            if (
                prev_type in [ImportType.QUICKSHELL, ImportType.QUICKSHELL_OTHER]
                and current_type == ImportType.QS_COMPONENTS
            ):
                result.append("\n")
            elif (
                current_type == ImportType.EXPLICIT and prev_type != ImportType.EXPLICIT
            ):
                result.append("\n")
        result.append(imp.line)
        prev_type = current_type
    if non_import_start > 0:
        result.append("\n")
    result.extend(lines[non_import_start:])

    # Check if the imports changed
    changed = [imp.line for imp in sorted_imports] != [imp.line for imp in imports]

    return result, changed


def classify_property(line: str) -> Optional[PropertyType]:
    stripped = line.strip()
    if re.match(r"property\s+alias\s+", stripped):
        return PropertyType.ALIAS
    elif re.match(r"required\s+property\s+", stripped):
        return PropertyType.REQUIRED
    elif re.match(r"readonly\s+property\s+", stripped):
        return PropertyType.READONLY
    elif re.match(r"property\s+\w+", stripped):
        return PropertyType.NORMAL
    return None


def find_component_structure(lines: List[str], start_idx: int) -> dict:
    structure = {
        "id": None,
        "anchors": [],
        "properties": [],
        "functions": [],
        "behaviors": [],
        "objects": [],
        "issues": [],
    }

    brace_count = 1
    i = start_idx + 1
    current_section = None
    base_indent = None

    while i < len(lines) and brace_count > 0:
        line = lines[i]
        stripped = line.strip()

        # Skip empty lines and comments
        if not stripped or stripped.startswith("//"):
            i += 1
            continue

        brace_count += line.count("{") - line.count("}")

        # Detect indentation level (process only top-level items)
        if base_indent is None and stripped:
            base_indent = len(line) - len(line.lstrip())

        current_indent = len(line) - len(line.lstrip())

        # Only process lines at base indentation level
        if base_indent is not None and current_indent == base_indent:
            # ID property
            if stripped.startswith("id:"):
                structure["id"] = (i, line)
            # Anchors
            elif stripped.startswith("anchors"):
                structure["anchors"].append((i, line))
            # Properties
            elif classify_property(line) is not None:
                prop_type = classify_property(line)
                structure["properties"].append((i, line, prop_type))
            # Functions
            elif re.match(r"function\s+\w+", stripped):
                structure["functions"].append((i, line))
            # Behaviors, Transitions, States
            elif re.match(r"(Behavior\s+on|Transition|States)\s*[:{[]", stripped):
                structure["behaviors"].append((i, line))
            # Objects (other components)
            elif re.match(r"[A-Z]\w+\s*\{", stripped):
                structure["objects"].append((i, line))

        i += 1

    return structure


def check_component_order(structure: dict) -> List[str]:
    issues = []

    if not structure["id"]:
        return issues

    id_line = structure["id"][0]
    last_line = id_line
    last_section = "id"

    # Expected order: id -> anchors -> properties -> functions -> behaviors -> objects
    sections = [
        ("anchors", structure["anchors"]),
        ("properties", structure["properties"]),
        ("functions", structure["functions"]),
        ("behaviors", structure["behaviors"]),
        ("objects", structure["objects"]),
    ]

    for section_name, items in sections:
        if not items:
            continue

        first_item_line = items[0][0]

        if first_item_line < last_line:
            issues.append(
                f"Line {first_item_line + 1}: {section_name} should be after {last_section}"
            )

        last_line = items[-1][0]
        last_section = section_name

    return issues


def check_property_order(properties: List[Tuple[int, str, PropertyType]]) -> List[str]:
    issues = []

    if len(properties) < 2:
        return issues

    prev_type = None
    for line_num, line, prop_type in properties:
        if prev_type is not None and prop_type.value < prev_type.value:
            issues.append(
                f"Line {line_num + 1}: Wrong property order. "
                f"Order: alias -> required -> readonly -> normal"
            )
        prev_type = prop_type

    return issues


def add_spacing_after_id(lines: List[str]) -> Tuple[List[str], bool]:
    result = []
    changed = False
    i = 0

    while i < len(lines):
        line = lines[i]
        stripped = line.strip()

        if stripped.startswith("id:"):
            result.append(line)

            next_idx = i + 1
            has_empty_line = False

            while next_idx < len(lines):
                next_line = lines[next_idx]
                next_stripped = next_line.strip()

                # If empty line is found, continue scanning
                if not next_stripped:
                    has_empty_line = True
                    result.append(next_line)
                    next_idx += 1
                    continue

                # Skip comments
                if next_stripped.startswith("//"):
                    result.append(next_line)
                    next_idx += 1
                    continue

                if next_stripped and next_stripped != "}":
                    if not has_empty_line:
                        result.append("\n")
                        changed = True
                break
            i = next_idx
            continue
        result.append(line)
        i += 1
    return result, changed


def format_qml_file(filepath: Path, check_only: bool = False) -> Tuple[bool, List[str]]:
    try:
        with open(filepath, "r", encoding="utf-8") as f:
            lines = f.readlines()
    except Exception as e:
        return False, [f"File read error: {e}"]

    issues = []
    changed = False

    new_lines, import_changed = sort_imports(lines)
    if import_changed:
        changed = True
        issues.append("✓ Fixed import order")

    if not check_only:
        new_lines, id_changed = add_spacing_after_id(new_lines)
        if id_changed:
            changed = True
            issues.append("✓ Added spacing after 'id'")

    for i, line in enumerate(new_lines):
        if re.match(r"^[A-Z]\w+\s*\{", line.strip()):
            structure = find_component_structure(new_lines, i)
            order_issues = check_component_order(structure)
            issues.extend(order_issues)
            if structure["properties"]:
                prop_issues = check_property_order(structure["properties"])
                issues.extend(prop_issues)

    if changed and not check_only:
        try:
            with open(filepath, "w", encoding="utf-8") as f:
                f.writelines(new_lines)
        except Exception as e:
            return False, [f"Error when writing file: {e}"]

    return changed or bool([i for i in issues if not i.startswith("✓")]), issues


def find_qml_files(directory: Path) -> List[Path]:
    """Find QML files using fd command for better performance."""
    try:
        result = subprocess.run(
            ["fd", "-e", "qml", "-t", "f", ".", str(directory)],
            capture_output=True,
            text=True,
            check=True,
        )
        files = [
            Path(line.strip())
            for line in result.stdout.strip().split("\n")
            if line.strip()
        ]
        return files
    except FileNotFoundError:
        # Fallback to Python's rglob if fd is not available
        print("Warning: 'fd' command not found, using slower fallback method")
        return list(directory.rglob("*.qml"))
    except subprocess.CalledProcessError:
        # Fallback on error
        return list(directory.rglob("*.qml"))


def run_custom_check(git_root: Path, verbose: bool = False) -> int:
    """Run custom QML check on all files."""
    print("\n" + "=" * 60)
    print("STEP 1: Running custom QML checks...")
    print("=" * 60)

    files = find_qml_files(git_root)
    if not files:
        print(f"No QML files found in {git_root}")
        return 0

    files_with_issues = 0
    total_manual_issues = 0

    for filepath in files:
        if verbose:
            print(f"\nProcessing: {filepath}")
        has_issues, issues = format_qml_file(filepath, check_only=True)
        if has_issues or issues:
            files_with_issues += 1
            print(f"\n{'='*60}")
            print(f"File: {filepath.relative_to(git_root)}")
            print("=" * 60)
            for issue in issues:
                print(f"  {issue}")
                if not issue.startswith("✓"):
                    total_manual_issues += 1

    print(f"\nCheck completed: {files_with_issues} of {len(files)} file(s) have issues")
    if total_manual_issues > 0:
        print(f"Manual fixes required: {total_manual_issues}")

    return total_manual_issues


def run_qmlformat(git_root: Path) -> int:
    """Run qmlformat on all QML files recursively."""
    print("\n" + "=" * 60)
    print("STEP 2: Running qmlformat...")
    print("=" * 60)

    files = find_qml_files(git_root)
    if not files:
        print("No QML files found")
        return 0

    failed_files = []
    for filepath in files:
        try:
            result = subprocess.run(
                ["qmlformat", "-i", str(filepath)],
                capture_output=True,
                text=True,
                cwd=git_root,
            )
            if result.returncode != 0:
                failed_files.append((filepath, result.stderr))
                print(f"✗ Failed: {filepath.relative_to(git_root)}")
                if result.stderr:
                    print(f"  Error: {result.stderr.strip()}")
            else:
                print(f"✓ Formatted: {filepath.relative_to(git_root)}")
        except FileNotFoundError:
            print("\nError: qmlformat not found. Please install Qt development tools.")
            return 1
        except Exception as e:
            failed_files.append((filepath, str(e)))
            print(f"✗ Error processing {filepath.relative_to(git_root)}: {e}")

    if failed_files:
        print(f"\nqmlformat failed on {len(failed_files)} file(s)")
        return 1
    else:
        print(f"\nqmlformat completed successfully on {len(files)} file(s)")
        return 0


def run_lupdate(git_root: Path) -> int:
    """Run lupdate to update translation files."""
    print("\n" + "=" * 60)
    print("STEP 3: Running lupdate...")
    print("=" * 60)

    translations_dir = git_root / "translations"

    if not translations_dir.exists():
        print(f"Warning: translations directory not found at {translations_dir}")
        print("Skipping lupdate step")
        return 0

    ts_files = list(translations_dir.glob("*.ts"))
    if not ts_files:
        print(f"Warning: No .ts files found in {translations_dir}")
        print("Skipping lupdate step")
        return 0

    try:
        result = subprocess.run(
            ["lupdate", ".", "-ts"] + [str(f) for f in ts_files],
            capture_output=True,
            text=True,
            cwd=git_root,
        )

        if result.returncode != 0:
            print(f"✗ lupdate failed")
            if result.stderr:
                print(f"Error: {result.stderr.strip()}")
            return 1
        else:
            print(f"✓ lupdate completed successfully")
            if result.stdout:
                print(result.stdout.strip())
            return 0
    except FileNotFoundError:
        print("\nError: lupdate not found. Please install Qt Linguist tools.")
        return 1
    except Exception as e:
        print(f"✗ Error running lupdate: {e}")
        return 1


def main():
    import argparse

    parser = argparse.ArgumentParser(
        description="QML formatter and checker - runs from git root"
    )
    parser.add_argument(
        "-v", "--verbose", action="store_true", help="Enable verbose logging"
    )
    parser.add_argument(
        "--skip-check", action="store_true", help="Skip custom QML check (step 1)"
    )
    parser.add_argument(
        "--skip-format", action="store_true", help="Skip qmlformat (step 2)"
    )
    parser.add_argument(
        "--skip-lupdate", action="store_true", help="Skip lupdate (step 3)"
    )

    args = parser.parse_args()

    # Find git root
    git_root = find_git_root()
    if git_root is None:
        print("Error: Not in a git repository")
        sys.exit(1)

    print(f"Git root: {git_root}")
    print(f"Current directory: {Path.cwd()}")

    # Change to git root
    os.chdir(git_root)

    exit_code = 0

    # Step 1: Custom check
    if not args.skip_check:
        check_result = run_custom_check(git_root, args.verbose)
        if check_result > 0:
            exit_code = 1

    # Step 2: qmlformat
    if not args.skip_format:
        format_result = run_qmlformat(git_root)
        if format_result != 0:
            exit_code = 1

    # Step 3: lupdate
    if not args.skip_lupdate:
        lupdate_result = run_lupdate(git_root)
        if lupdate_result != 0:
            exit_code = 1

    # Final summary
    print("\n" + "=" * 60)
    print("SUMMARY")
    print("=" * 60)
    if exit_code == 0:
        print("✓ All steps completed successfully")
    else:
        print("✗ Some steps failed or require manual fixes")

    sys.exit(exit_code)


if __name__ == "__main__":
    main()

File: Assets/go/formatting.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

type ImportType int

const (
	QT_CORE ImportType = iota + 1
	QT_QUICK
	QT_LABS
	QT_QUICK_OTHER
	QUICKSHELL
	QUICKSHELL_OTHER
	QS_COMPONENTS
	EXPLICIT
)

type PropertyType int

const (
	ALIAS PropertyType = iota + 1
	REQUIRED
	READONLY
	NORMAL
)

type Import struct {
	Line          string
	Type          ImportType
	OriginalIndex int
}

type Property struct {
	Line   string
	Type   PropertyType
	Indent string
}

type ComponentStructure struct {
	ID           *LineInfo
	Anchors      []LineInfo
	Properties   []PropertyInfo
	Functions    []LineInfo
	Behaviors    []LineInfo
	Objects      []LineInfo
	NestedStarts []int
}

type LineInfo struct {
	Index int
	Line  string
}

type PropertyInfo struct {
	Index string
	Line  string
	Type  PropertyType
}

func findGitRoot() (string, error) {
	out, err := exec.Command("git", "rev-parse", "--show-toplevel").Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}

func classifyImport(line string) ImportType {
	re := regexp.MustCompile(`import\s+([^\s]+)`)
	matches := re.FindStringSubmatch(line)
	if len(matches) < 2 {
		return EXPLICIT
	}

	module := matches[1]
	if strings.HasPrefix(module, `"`) || strings.HasPrefix(module, `'`) {
		return EXPLICIT
	}

	switch {
	case module == "QtCore":
		return QT_CORE
	case module == "QtQuick":
		return QT_QUICK
	case strings.HasPrefix(module, "Qt.labs."):
		return QT_LABS
	case strings.HasPrefix(module, "QtQuick."):
		return QT_QUICK_OTHER
	case module == "Quickshell":
		return QUICKSHELL
	case strings.HasPrefix(module, "Quickshell."):
		return QUICKSHELL_OTHER
	case strings.HasPrefix(module, "qs."):
		return QS_COMPONENTS
	default:
		return EXPLICIT
	}
}

func sortImports(lines []string) ([]string, bool) {
	var imports []Import
	nonImportStart := -1

	for i, line := range lines {
		stripped := strings.TrimSpace(line)
		if strings.HasPrefix(stripped, "import ") {
			importType := classifyImport(stripped)
			imports = append(imports, Import{Line: line, Type: importType, OriginalIndex: i})
		} else if stripped != "" && !strings.HasPrefix(stripped, "//") {
			nonImportStart = i
			break
		}
	}

	if len(imports) == 0 {
		return lines, false
	}

	sortedImports := make([]Import, len(imports))
	copy(sortedImports, imports)
	sort.SliceStable(sortedImports, func(i, j int) bool {
		return sortedImports[i].Type < sortedImports[j].Type
	})

	var result []string
	var prevType ImportType = 0
	changed := false

	for i, imp := range sortedImports {
		if i < len(imports) && imp.Line != imports[i].Line {
			changed = true
		}

		if prevType != 0 {
			if (prevType == QUICKSHELL || prevType == QUICKSHELL_OTHER) && imp.Type == QS_COMPONENTS {
				result = append(result, "\n")
			} else if imp.Type == EXPLICIT && prevType != EXPLICIT {
				result = append(result, "\n")
			}
		}
		result = append(result, imp.Line)
		prevType = imp.Type
	}

	if nonImportStart > 0 {
		result = append(result, "\n")
		result = append(result, lines[nonImportStart:]...)
	} else if nonImportStart == -1 {
		// Only imports or empty/comments
	} else if nonImportStart == 0 {
		// Should not happen if imports exist
		result = append(result, lines...)
	}

	return result, changed
}

func classifyProperty(line string) *PropertyType {
	stripped := strings.TrimSpace(line)
	var pt PropertyType
	if regexp.MustCompile(`^property\s+alias\s+`).MatchString(stripped) {
		pt = ALIAS
	} else if regexp.MustCompile(`^required\s+property\s+`).MatchString(stripped) {
		pt = REQUIRED
	} else if regexp.MustCompile(`^readonly\s+property\s+`).MatchString(stripped) {
		pt = READONLY
	} else if regexp.MustCompile(`^property\s+\w+`).MatchString(stripped) {
		pt = NORMAL
	} else {
		return nil
	}
	return &pt
}

func findComponentStructure(lines []string, startIdx int) ComponentStructure {
	structure := ComponentStructure{}
	braceCount := 1
	i := startIdx + 1
	baseIndent := -1

	for i < len(lines) && braceCount > 0 {
		line := lines[i]
		stripped := strings.TrimSpace(line)

		if stripped == "" || strings.HasPrefix(stripped, "//") {
			i++
			continue
		}

		openBraces := strings.Count(line, "{")
		closeBraces := strings.Count(line, "}")
		braceCount += openBraces - closeBraces

		currentIndent := len(line) - len(strings.TrimLeft(line, " \t"))
		if baseIndent == -1 && stripped != "" {
			baseIndent = currentIndent
		}

		if baseIndent != -1 && currentIndent == baseIndent {
			if strings.HasPrefix(stripped, "id:") {
				structure.ID = &LineInfo{Index: i, Line: line}
			} else if strings.HasPrefix(stripped, "anchors") {
				structure.Anchors = append(structure.Anchors, LineInfo{Index: i, Line: line})
			} else if pt := classifyProperty(line); pt != nil {
				structure.Properties = append(structure.Properties, PropertyInfo{Index: fmt.Sprintf("%d", i), Line: line, Type: *pt})
			} else if regexp.MustCompile(`^function\s+\w+`).MatchString(stripped) {
				structure.Functions = append(structure.Functions, LineInfo{Index: i, Line: line})
			} else if regexp.MustCompile(`^(Behavior\s+on|Transition|States)\s*[:{[]`).MatchString(stripped) {
				structure.Behaviors = append(structure.Behaviors, LineInfo{Index: i, Line: line})
			} else if regexp.MustCompile(`^[A-Z]\w+\s*\{`).MatchString(stripped) {
				structure.Objects = append(structure.Objects, LineInfo{Index: i, Line: line})
				structure.NestedStarts = append(structure.NestedStarts, i) // <-- record for recursion
			}
		} else if baseIndent != -1 && currentIndent > baseIndent {
			if regexp.MustCompile(`^[A-Z]\w+\s*\{`).MatchString(stripped) {
				structure.NestedStarts = append(structure.NestedStarts, i)
			}
		}

		i++
	}
	return structure
}

func collectComponentIssues(lines []string, startIdx int) []string {
	var issues []string
	structure := findComponentStructure(lines, startIdx)

	issues = append(issues, checkComponentOrder(structure)...)
	if len(structure.Properties) > 0 {
		issues = append(issues, checkPropertyOrder(structure.Properties)...)
	}

	for _, nestedIdx := range structure.NestedStarts {
		issues = append(issues, collectComponentIssues(lines, nestedIdx)...)
	}

	return issues
}

func checkComponentOrder(structure ComponentStructure) []string {
	var issues []string
	if structure.ID == nil {
		return issues
	}

	lastLine := structure.ID.Index
	lastSection := "id"

	type section struct {
		name  string
		items []LineInfo
	}

	sections := []section{
		{"anchors", structure.Anchors},
		{"properties", propertyInfoToLineInfo(structure.Properties)},
		{"functions", structure.Functions},
		{"behaviors", structure.Behaviors},
		{"objects", structure.Objects},
	}

	for _, s := range sections {
		if len(s.items) == 0 {
			continue
		}

		firstItemLine := s.items[0].Index
		if firstItemLine < lastLine {
			issues = append(issues, fmt.Sprintf("Line %d: %s should be after %s", firstItemLine+1, s.name, lastSection))
		}

		lastLine = s.items[len(s.items)-1].Index
		lastSection = s.name
	}

	return issues
}

func propertyInfoToLineInfo(pi []PropertyInfo) []LineInfo {
	var li []LineInfo
	for _, p := range pi {
		var idx int
		fmt.Sscanf(p.Index, "%d", &idx)
		li = append(li, LineInfo{Index: idx, Line: p.Line})
	}
	return li
}

func checkPropertyOrder(properties []PropertyInfo) []string {
	var issues []string
	if len(properties) < 2 {
		return issues
	}

	var prevType PropertyType = 0
	for _, p := range properties {
		if prevType != 0 && p.Type < prevType {
			var idx int
			fmt.Sscanf(p.Index, "%d", &idx)
			issues = append(issues, fmt.Sprintf("Line %d: Wrong property order. Order: alias -> required -> readonly -> normal", idx+1))
		}
		prevType = p.Type
	}
	return issues
}

func addSpacingAfterID(lines []string) ([]string, bool) {
	var result []string
	changed := false
	i := 0
	for i < len(lines) {
		line := lines[i]
		stripped := strings.TrimSpace(line)

		if strings.HasPrefix(stripped, "id:") {
			result = append(result, line)
			nextIdx := i + 1
			hasEmptyLine := false

			for nextIdx < len(lines) {
				nextLine := lines[nextIdx]
				nextStripped := strings.TrimSpace(nextLine)

				if nextStripped == "" {
					hasEmptyLine = true
					result = append(result, nextLine)
					nextIdx++
					continue
				}

				if strings.HasPrefix(nextStripped, "//") {
					result = append(result, nextLine)
					nextIdx++
					continue
				}

				if nextStripped != "" && nextStripped != "}" {
					if !hasEmptyLine {
						result = append(result, "\n")
						changed = true
					}
				}
				break
			}
			i = nextIdx
			continue
		}
		result = append(result, line)
		i++
	}
	return result, changed
}

func formatQMLFile(filepath string, checkOnly bool) (bool, []string) {
	content, err := os.ReadFile(filepath)
	if err != nil {
		return false, []string{fmt.Sprintf("File read error: %v", err)}
	}

	lines := strings.Split(string(content), "\n")
	// Ensure last empty line is handled correctly if it exists
	if len(lines) > 0 && lines[len(lines)-1] == "" && !strings.HasSuffix(string(content), "\n") {
		// do nothing
	}

	issues := []string{}
	changed := false

	newLines, importChanged := sortImports(lines)
	if importChanged {
		changed = true
		issues = append(issues, "✓ Fixed import order")
	}

	if !checkOnly {
		_, idChanged := addSpacingAfterID(newLines)
		if idChanged {
			changed = true
			issues = append(issues, "✓ Added spacing after 'id'")
		}
	}

	for i, line := range newLines {
		if regexp.MustCompile(`^[A-Z]\w+\s*\{`).MatchString(strings.TrimSpace(line)) {
			issues = append(issues, collectComponentIssues(newLines, i)...)
			break
		}
	}

	if changed && !checkOnly {
		output := strings.Join(newLines, "\n")
		err := os.WriteFile(filepath, []byte(output), 0644)
		if err != nil {
			return false, []string{fmt.Sprintf("Error when writing file: %v", err)}
		}
	}

	manualIssues := []string{}
	for _, issue := range issues {
		if !strings.HasPrefix(issue, "✓") {
			manualIssues = append(manualIssues, issue)
		}
	}

	return len(manualIssues) > 0, issues
}

func findQMLFiles(directory string) []string {
	out, err := exec.Command("fd", "-e", "qml", "-t", "f", ".", directory).Output()
	if err != nil {
		// Fallback to manual walk
		var files []string
		filepath.Walk(directory, func(path string, info os.FileInfo, err error) error {
			if err == nil && !info.IsDir() && strings.HasSuffix(path, ".qml") {
				files = append(files, path)
			}
			return nil
		})
		return files
	}

	lines := strings.Split(strings.TrimSpace(string(out)), "\n")
	var files []string
	for _, line := range lines {
		if line != "" {
			files = append(files, line)
		}
	}
	return files
}

func runCustomCheck(gitRoot string, verbose bool) int {
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 1: Running custom QML checks...")
	fmt.Println(strings.Repeat("=", 60))

	files := findQMLFiles(gitRoot)
	if len(files) == 0 {
		fmt.Printf("No QML files found in %s\n", gitRoot)
		return 0
	}

	filesWithIssues := 0
	totalManualIssues := 0

	for _, fp := range files {
		if verbose {
			fmt.Printf("\nProcessing: %s\n", fp)
		}
		hasIssues, issues := formatQMLFile(fp, true)
		if hasIssues {
			filesWithIssues++
			rel, _ := filepath.Rel(gitRoot, fp)
			fmt.Println("\n" + strings.Repeat("=", 60))
			fmt.Printf("File: %s\n", rel)
			fmt.Println(strings.Repeat("=", 60))
			for _, issue := range issues {
				fmt.Printf("  %s\n", issue)
				if !strings.HasPrefix(issue, "✓") {
					totalManualIssues++
				}
			}
		}
	}

	fmt.Printf("\nCheck completed: %d of %d file(s) have issues\n", filesWithIssues, len(files))
	if totalManualIssues > 0 {
		fmt.Printf("Manual fixes required: %d\n", totalManualIssues)
	}
	return totalManualIssues
}

func runQmlformat(gitRoot string) int {
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 2: Running qmlformat...")
	fmt.Println(strings.Repeat("=", 60))

	files := findQMLFiles(gitRoot)
	if len(files) == 0 {
		fmt.Println("No QML files found")
		return 0
	}

	failedFiles := 0
	for _, fp := range files {
		cmd := exec.Command("qmlformat", "-i", fp)
		cmd.Dir = gitRoot
		out, err := cmd.CombinedOutput()
		if err != nil {
			failedFiles++
			rel, _ := filepath.Rel(gitRoot, fp)
			fmt.Printf("✗ Failed: %s\n", rel)
			fmt.Printf("  Error: %s\n", strings.TrimSpace(string(out)))
		}
	}

	if failedFiles > 0 {
		fmt.Printf("\nqmlformat failed on %d file(s)\n", failedFiles)
		return 1
	}
	fmt.Printf("\nqmlformat completed successfully on %d file(s)\n", len(files))
	return 0
}

func runLupdate(gitRoot string) int {
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 3: Running lupdate...")
	fmt.Println(strings.Repeat("=", 60))

	translationsDir := filepath.Join(gitRoot, "translations")
	if _, err := os.Stat(translationsDir); os.IsNotExist(err) {
		fmt.Printf("Warning: translations directory not found at %s\n", translationsDir)
		return 0
	}

	entries, _ := os.ReadDir(translationsDir)
	var tsFiles []string
	for _, entry := range entries {
		if !entry.IsDir() && strings.HasSuffix(entry.Name(), ".ts") {
			tsFiles = append(tsFiles, filepath.Join(translationsDir, entry.Name()))
		}
	}

	if len(tsFiles) == 0 {
		fmt.Printf("Warning: No .ts files found in %s\n", translationsDir)
		return 0
	}

	args := append([]string{".", "-ts"}, tsFiles...)
	cmd := exec.Command("lupdate", args...)
	cmd.Dir = gitRoot
	out, err := cmd.CombinedOutput()

	if err != nil {
		fmt.Println("✗ lupdate failed")
		fmt.Printf("Error: %s\n", strings.TrimSpace(string(out)))
		return 1
	}

	fmt.Println("✓ lupdate completed successfully")
	if len(out) > 0 {
		fmt.Println(strings.TrimSpace(string(out)))
	}
	return 0
}

func main() {
	verbose := false
	skipCheck := false
	skipFormat := false
	skipLupdate := false

	for _, arg := range os.Args[1:] {
		switch arg {
		case "-v", "--verbose":
			verbose = true
		case "--skip-check":
			skipCheck = true
		case "--skip-format":
			skipFormat = true
		case "--skip-lupdate":
			skipLupdate = true
		}
	}

	gitRoot, err := findGitRoot()
	if err != nil {
		fmt.Println("Error: Not in a git repository")
		os.Exit(1)
	}

	fmt.Printf("Git root: %s\n", gitRoot)
	cwd, _ := os.Getwd()
	fmt.Printf("Current directory: %s\n", cwd)

	os.Chdir(gitRoot)
	exitCode := 0

	if !skipCheck {
		if runCustomCheck(gitRoot, verbose) > 0 {
			exitCode = 1
		}
	}

	if !skipFormat {
		if runQmlformat(gitRoot) != 0 {
			exitCode = 1
		}
	}

	if !skipLupdate {
		if runLupdate(gitRoot) != 0 {
			exitCode = 1
		}
	}

	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("SUMMARY")
	fmt.Println(strings.Repeat("=", 60))
	if exitCode == 0 {
		fmt.Println("✓ All steps completed successfully")
	} else {
		fmt.Println("✗ Some steps failed or require manual fixes")
	}

	os.Exit(exitCode)
}
